src/main.js

```diff
@@ -129,48 +140,241 @@ const drawBackground = () => {
+          ],
+          category: "image",
+          creators: solanaMetadata.creators,
+        },
+      };
+      break;
+    }
   }
+
   metadataList.push(tempMetadata);
   attributesList = [];
 };
 
+const addRarityMetadata = () => {
+  let traitOccurances = [];
+  let totalAttributesCnt = 0;
+
+  // count occurrences for all traits/layers & attributes/assets
+  metadataList.forEach((item) => {
+    item.attributes.forEach((a) => {
+      if (rarityObject[a.trait_type] != null) {
+        if (rarityObject[a.trait_type][a.value] != null) {
+          rarityObject[a.trait_type][a.value].attributeOccurrence++;
+        } else {
+          rarityObject[a.trait_type][a.value] = { attributeOccurrence: 1 };
+          totalAttributesCnt++;
+        }
+
+        traitOccurances[a.trait_type]++;
+      } else {
+        rarityObject[a.trait_type] = { [a.value]: { attributeOccurrence: 1 } };
+        traitOccurances[a.trait_type] = 1;
+        totalAttributesCnt++;
+      }
+    });
+  });
+
+  // general rarity metadata for all traits/layers & attributes/assets
+  Object.entries(rarityObject).forEach((entry) => {
+    const layer = entry[0];
+    const assets = entry[1];
+
+    Object.entries(assets).forEach((asset) => {
+      // trait/layer related
+      asset[1].traitOccurance = traitOccurances[layer];
+      asset[1].traitOccurancePercentage =
+        (metadataList.length / asset[1].traitOccurance) * 100;
+
+      // attribute/asset related
+      asset[1].attributeOccurrencePercentage =
+        (asset[1].attributeOccurrence / metadataList.length) * 100;
+
+      // TR / SR algorithms specific metadata
+      if (
+        network.rarityAlgorithm == rarityAlgorithms.TraitRarity ||
+        network.rarityAlgorithm == rarityAlgorithms.StatisticalRarity ||
+        network.rarityAlgorithm == rarityAlgorithms.TraitAndStatisticalRarity
+      ) {
+        // logic from https://github.com/xterr/nft-generator/blob/d8992d2bcfa729a6b2ef443f9404ffa28102111b/src/components/RarityResolver.ts
+        // ps: the only difference being that attributeRarityNormed is calculated only once
+        const totalLayersCnt = Object.keys(traitOccurances).length;
+        const avgAttributesPerTrait = totalAttributesCnt / totalLayersCnt;
+        asset[1].attributeFrequency =
+          asset[1].attributeOccurrence / metadataList.length;
+        asset[1].traitFrequency = asset[1].traitOccurance > 0 ? 1 : 0;
+        asset[1].attributeRarity =
+          metadataList.length / asset[1].attributeOccurrence;
+        asset[1].attributeRarityNormed =
+          asset[1].attributeRarity * (avgAttributesPerTrait / totalLayersCnt);
+      }
+    });
+  });
+
+  // calculate rarity for each item/NFT
+  if (network.rarityAlgorithm == rarityAlgorithms.none) {
+    return;
+  } else if (network.rarityAlgorithm == rarityAlgorithms.JaccardDistances) {
+    let z = [];
+    let avg = [];
+
+    // calculate z(i,j) and avg(i)
+    for (let i = 0; i < metadataList.length; i++) {
```

Comment:
Could you please provide the docs that you used for the implementation? JD usually is between 0-1, I see different results. Is this just the actual JD*100?

Reply:
It's just JD*100
https://www.publish0x.com/navid-ladani/explaining-nft-rarity-score-models-xmmorpr